### PR TITLE
Datetimeinterface

### LINKED ISSUE #1710: DateTimes should be type-hinted as DateTimeInterface
## Feature Request

Currently, properties containing a date (or a datetime) are type-hintend in phpdoc with `DateTime`.
Since Doctrine also allows to map properties with `DateTimeImmutable`, such properties should be type-hinted with `DateTimeInterface`.

See for example https://github.com/sonata-project/SonataMediaBundle/blob/3.x/src/Model/Media.php#L103

### PR DESCRIPTION
**Pull request by garak**
## Subject

As detailed in related issue, this PR replace type hints and return type of DateTime with broader DateTimeInterface, allowing to use DateTimeImmutable.

I am targeting this branch, because change is BC.

Closes #1710

## Changelog

Replaced DateTime type hints and return types with DateTimeInterface 

```markdown
### Changed
- Replaced DateTime type hints and return types with DateTimeInterface 
```

Added some checks in testing, to skip tests when not appliable.
Notice that this can be fixed in some other ways (e.g. by forcing some requires in dev). Anyway, I couldn't run my tests without such changes.

### CODE REVIEW

**Comment by VincentLanglet** [comment]
There is something wrong with your branch, it has a lot of commit

**Comment by garak** [comment]
> There is something wrong with your branch, it has a lot of commit

I noticed that, and tried to rebase my branch on 3.x, but no luck.
Any suggestions?

**Comment by VincentLanglet** [comment]
> Any suggestions?

```
git reset --hard origin/3.x
git cherry-pick 4df4fb9
```

File: tests/CDN/PantherPortalTest.php [hunk 1]
```diff
@@ -20,6 +20,10 @@ class PantherPortalTest extends TestCase
 {
     public function testPortal(): void
```
Comment (OskarStark) [inline tests/CDN/PantherPortalTest.php:21]: Symfony uses @require annotation can we use this too? 

Also this should be fixed in a separate PR

File: tests/Form/Type/MediaTypeTest.php [hunk 2]
```diff
@@ -67,6 +67,10 @@ public function testMissingFormOptions(): void
 
     public function testMissingFormContextOption(): void
     {
+        if (!\function_exists('gd_info')) {
```
Comment (OskarStark) [inline tests/Form/Type/MediaTypeTest.php:70]: Same here for the @require annotation

**Review by OskarStark — CHANGES_REQUESTED** [review]

File: tests/CDN/PantherPortalTest.php (hunk 1 above)
Comment (garak) [reply tests/CDN/PantherPortalTest.php:21]: Never heard about it, can you point me to an example?

Comment (OskarStark) [reply tests/CDN/PantherPortalTest.php:21]: Sure, I am currently on a phone 📱, will have a look later

File: src/Model/Gallery.php [hunk 3]
```diff
@@ -82,7 +82,7 @@ public function getEnabled()
         return $this->enabled;
     }
 
-    public function setUpdatedAt(?\DateTime $updatedAt = null)
+    public function setUpdatedAt(?\DateTimeInterface $updatedAt = null)
```
Comment (OskarStark) [inline src/Model/Gallery.php:85]: This is a BC break isn't it? 🤔

Comment (garak) [reply src/Model/Gallery.php:85]: Why?

Comment (VincentLanglet) [reply src/Model/Gallery.php:85]: I think it is, but it feel so bad to delay this PR to a major version...
Maybe an upgrade note could be enough

Comment (VincentLanglet) [reply src/Model/Gallery.php:85]: @garak In case someone was overriding the function.

```
class A {
    public function foo(\DatetimeInterface $a) {}
}

class B extends A {
    public function foo(\Datetime $a) {}
}
```
is not valid

File: tests/CDN/PantherPortalTest.php (hunk 1 above)
Comment (garak) [reply tests/CDN/PantherPortalTest.php:21]: never mind, found it (and added it)

Comment (OskarStark) [reply tests/CDN/PantherPortalTest.php:21]: Does it work?

File: tests/CDN/PantherPortalTest.php [hunk 4]
```diff
@@ -18,6 +18,9 @@
 
 class PantherPortalTest extends TestCase
 {
+    /**
+     * @requires function \SoapClient::__construct
```
Comment (OskarStark) [inline tests/CDN/PantherPortalTest.php:22]:
```suggestion
     * @requires extension soap
```

File: tests/CDN/PantherPortalTest.php [hunk 5]
```diff
@@ -35,6 +38,9 @@ public function testPortal(): void
         $panther->flushPaths([$path]);
     }
 
+    /**
+     * @requires function \SoapClient::__construct
```
Comment (OskarStark) [inline tests/CDN/PantherPortalTest.php:42]:
```suggestion
     * @requires extension soap
```

File: tests/CDN/PantherPortalTest.php [hunk 6]
```diff
@@ -50,10 +56,12 @@ public function testException(): void
     }
 }
 
-class ClientSpy extends \SoapClient
```
Comment (OskarStark) [inline tests/CDN/PantherPortalTest.php:53]: I don't understand why we have a class, if we mock it anyway, can't we directly mock it or better create an anonymous class:
```php
        $client = new class() implements \SoapClient {
            public function flush(): string
            {
                return 'hello';
            }
        };
```

WDYT @VincentLanglet @franmomu ?

File: tests/CDN/PantherPortalTest.php [hunk 7]
```diff
@@ -50,10 +56,12 @@ public function testException(): void
...
-{
-    public function flush(): string
+if (class_exists(\SoapClient::class)) {
+    class ClientSpy extends \SoapClient
```
Comment (OskarStark) [inline tests/CDN/PantherPortalTest.php:60]: Can we remove the if class exists check if we move it to an extra file? 🤔

File: tests/Form/Type/MediaTypeTest.php [hunk 8]
```diff
@@ -123,6 +126,9 @@ public function testInvalidFormProviderOption(): void
         ]);
     }
 
+    /**
+     * @requires extension gd_info
```
Comment (OskarStark) [inline tests/Form/Type/MediaTypeTest.php:130]: IMO the extension is `gd`:
```suggestion
     * @requires extension gd
```

File: tests/Functional/Controller/TruncateControllerTest.php [hunk 9]
```diff
@@ -21,6 +21,9 @@
 
 final class TruncateControllerTest extends TestCase
 {
+    /**
+     * @requires extension gd_info
```
Comment (OskarStark) [inline tests/Functional/Controller/TruncateControllerTest.php:25]: IMO the extension is `gd`:
```suggestion
     * @requires extension gd
```

File: src/Model/Gallery.php (hunk 3 above)
Comment (OskarStark) [reply src/Model/Gallery.php:85]: But we can remove the typehint now and trigger an exception like:

/**
 * NEXT_MAJOR: Add the typehint back.
 */
public function setUpdatedAt(/* ?\DateTimeInterface */$updatedAt = null)
{
    trigger deprecation here to warn the user it needs to be an DateTimeInterface?
    // ...
}

Comment (garak) [reply src/Model/Gallery.php:85]: @OskarStark but how can you check for argument? You can't check for interfaces

**Review by VincentLanglet — CHANGES_REQUESTED** [review]
Can you add a line in https://github.com/sonata-project/SonataMediaBundle/blob/3.x/UPGRADE-3.x.md#upgrade-from-3x-to-3x to explain about the little BC-break ?

File: UPGRADE-3.x.md [hunk 10]
```diff
@@ -43,6 +43,10 @@ are required when aws/aws-sdk-php 3.x is installed.
 
 The `Guzzle` and `Buzz` dependencies are deprecated and will be replaced with the abstract `http-client` interface, so you can choose your preferred client implementation. You should adapt to the new `BaseVideoProvider::__construct()` signature.
 
+## Model classes use `\DateTimeInterface`
+
+If you overrided `setUpdatedAt` method in model classes (Gallery, Media, GalleryHasMedia), you need to change type hint of argument to `\DateTimeInterface`.
```
Comment (VincentLanglet) [inline UPGRADE-3.x.md:48]: There is also `setCdnFlushAt` and `setCreatedAt`

**Review by VincentLanglet — APPROVED** [review]

File: src/Model/Gallery.php (hunk 3 above)
Comment (franmomu) [reply src/Model/Gallery.php:85]: What do you mean @garak? IMO we should not do this (as it is right now, introducing a BC break) in the stable branch, these models are meant to be extended by the user.

Comment (VincentLanglet) [reply src/Model/Gallery.php:85]:
```
class A {
    public function foo($a) {}
}

class B extends A {
    public function foo(\Datetime $a) {}
}
```
Is not more valid IMHO @OskarStark 

> IMO we should not do this (as it is right now, introducing a BC break) in the stable branch, these models are meant to be extended by the user.

Model are meant to be extended but these setters shouldn't be a lot

Comment (garak) [reply src/Model/Gallery.php:85]: @franmomu I think we can safely change since likely no one is overriding such methods.
I can't see a feasible way to check for argument, in the case we remove the type hint.

Comment (core23) [reply src/Model/Gallery.php:85]: This change will break my project(s)!

> Model are meant to be extended but these setters shouldn't be a lot

For a greenfield project, you're right, This model classes were designed for extensions, (e.g. the entity class will do this).

**Review by core23 — DISMISSED** [review]

File: src/Model/Gallery.php (hunk 3 above)
Comment (OskarStark) [reply src/Model/Gallery.php:85]: It is, but since 7.4: https://3v4l.org/iCa85

Comment (VincentLanglet) [reply src/Model/Gallery.php:85]: So should we do this PR on master ? :/

Comment (core23) [reply src/Model/Gallery.php:85]: Yes, that's the only way :(

**Comment by VincentLanglet** [comment]
Can this PR be done on master ? :)

**Comment by VincentLanglet** [comment]
@jordisala1991 recently worked on the version 4 of this bundle.
@garak Do you have time to make this PR on master ?

I've added the 4.0 milestone in order to not forget to take this BC-break.

**Comment by garak** [comment]
> @garak Do you have time to make this PR on master ?

Done

**Review by OskarStark — APPROVED** [review]

**Comment by VincentLanglet** [comment]
There is some conflicts

**Comment by garak** [comment]
> There is some conflicts

I just rebased

File: UPGRADE-3.x.md [hunk 11]
```diff
@@ -51,6 +51,11 @@ are required when aws/aws-sdk-php 3.x is installed.
 
 The `Guzzle` and `Buzz` dependencies are deprecated and will be replaced with the abstract `http-client` interface, so you can choose your preferred client implementation. You should adapt to the new `BaseVideoProvider::__construct()` signature.
 
+## Model classes use `\DateTimeInterface`
```
[the remark this answers is not in the document]
Comment (garak) [reply UPGRADE-3.x.md]: Moved to UGRADE-4.0md

**Review by VincentLanglet — APPROVED** [review]

File: UPGRADE-4.0.md [hunk 12]
```diff
@@ -49,6 +49,9 @@ If you have implemented a custom model, you must adapt the signature of the foll
  * `GalleryHasMediaInterface::getId`
  * `GalleryInterface::getId`
 
+If you overrided some date-related methods (`setUpdatedAt`, `setCreatedAt`, `setCdnFlushAt`) in model classes (Gallery, Media, GalleryHasMedia),
+you need to change type hint of argument to `\DateTimeInterface`.
```
Comment (phansys) [inline UPGRADE-4.0.md:53]:
```suggestion
If you have overridden some date-related methods (`setUpdatedAt()`, `setCreatedAt()`, `setCdnFlushAt()`) in model classes (`Gallery`, `Media`, `GalleryHasMedia`),
you need to change the arguments' type declarations to `\DateTimeInterface`.
```

**Review by phansys — APPROVED** [review]

**Review by jordisala1991 — APPROVED** [review]

**Comment by jordisala1991** [comment]
Thank you @garak
